### PR TITLE
Add linter to the scss themebuilder

### PR DESCRIPTION
**Pull request by babich-a**

### CODE REVIEW

File: themebuilder-scss/.eslintignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+coverage
+dist
+node_modules
```
Comment (San4es) [inline themebuilder-scss/.eslintignore]: https://eslint.org/docs/user-guide/configuring#eslintignore
> In addition to any patterns in a .eslintignore file, ESLint ignores files in /**/node_modules/* by default.
```suggestion

```

File: .eslintignore [hunk 2]
```diff
@@ -2,4 +2,5 @@ artifacts/*
 js/viz/docs/*
 node_modules/*
 testing/helpers/sinon/*
-/themebuilder/data/metadata/*
+themebuilder/data/metadata/*
+themebuilder-scss/**/*
```
Comment (San4es) [inline .eslintignore:6]: https://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy
>To limit ESLint to a specific project, place "root": true inside the eslintConfig field of the package.json file or in the .eslintrc.* file at your project's root level. ESLint will stop looking in parent folders once it finds a configuration with "root": true.

So, let's add `"root": true ` to [ThemeBuilder config](https://github.com/DevExpress/DevExtreme/pull/13036/files#diff-cf5d90eb805257fe74ea00764ceb6771)?

File: themebuilder-scss/package.json [hunk 3]
```diff
@@ -19,7 +19,9 @@
     "semver": "^5.6.0"
   },
   "scripts": {
-    "test": "jest --coverage --verbose --detectOpenHandles",
+    "test-code": "jest --coverage --verbose --detectOpenHandles",
+    "lint": "eslint --config .eslintrc-themebuilder .",
```
Comment (San4es) [inline themebuilder-scss/package.json:23]: It looks like `"root": true` helps us to simplify this command to:
```"lint": "eslint ."```  See https://github.com/DevExpress/DevExtreme/pull/13036#discussion_r427818376

File: docker-ci.sh [hunk 4]
```diff
@@ -193,6 +193,9 @@ function run_test_scss {
     npx gulp generate-scss
     npm run build-themes
     node build/gulp/scss/tests/identical.test.js
+
+    cd themebuilder-scss
+    npm i && npm run build && npm run test
```
Comment (San4es) [inline docker-ci.sh:198]: What do you think about moving the `lint` script call here?
https://github.com/DevExpress/DevExtreme/blob/8794fe2189f5c419b6f4a6840f6a2689e2ea07f6/docker-ci.sh#L10-L13

File: themebuilder-scss/.eslintrc-themebuilder [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{
+    "env": {
+        "es6": true,
+        "node": true
+    },
+    "parser": "@typescript-eslint/parser",
+    "plugins": [
+        "@typescript-eslint"
```
Comment (San4es) [inline themebuilder-scss/.eslintrc-themebuilder:8]: What about [eslint-plugin-node](https://www.npmjs.com/package/eslint-plugin-node)?

File: themebuilder-scss/package.json [hunk 6]
```diff
@@ -29,13 +31,20 @@
     "@types/clean-css": "^4.2.1",
     "@types/jest": "^25.2.1",
     "@types/node-sass": "^4.11.0",
-    "eslint": "^6.8.0",
+    "@typescript-eslint/eslint-plugin": "^2.33.0",
+    "eslint": "^7.0.0",
+    "eslint-config-airbnb-typescript": "^7.2.1",
```
Comment (San4es) [inline themebuilder-scss/package.json:36]: Why do we need this package?

File: themebuilder-scss/src/data/bootstrap-metadata/bootstrap-metadata.ts [hunk 7]
```diff
@@ -1,66 +1,66 @@
 export default {
-    'base-font-family': '@font-family-sans-serif',
-    'base-accent': '@brand-primary',
-    'base-bg': '@body-bg',
-    'base-text-color': '@text-color',
-    'base-border-color': '@table-border-color',
+  'base-font-family': '@font-family-sans-serif',
```
Comment (San4es) [inline themebuilder-scss/src/data/bootstrap-metadata/bootstrap-metadata.ts:2]: Now we use 4 spaces for indents everywhere.

File: themebuilder-scss/.eslintrc-themebuilder [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    ],
+    "parserOptions": {
+        "project": "./tsconfig.json",
+        "ecmaVersion": 6,
+        "sourceType": "module"
+    },
+    "extends": [
+        "devextreme/typescript"
```
Comment (San4es) [inline themebuilder-scss/.eslintrc-themebuilder:16]: Let's try to add also:
```
"eslint:recommended"
"plugin:@typescript-eslint/eslint-recommended"
```
?

See: 
https://www.npmjs.com/package/@typescript-eslint/eslint-plugin#recommended-configs

File: themebuilder-scss/package.json [hunk 9]
```diff
@@ -29,13 +31,20 @@
...
+    "eslint-config-devextreme": "^0.1.0",
+    "eslint-plugin-import": "^2.20.2",
+    "eslint-plugin-jest": "^23.11.0",
+    "eslint-plugin-jest-formatting": "^1.2.0",
+    "eslint-plugin-jsx-a11y": "^6.2.3",
+    "eslint-plugin-react": "^7.20.0",
```
Comment (San4es) [inline themebuilder-scss/package.json:42]: @churkin  can we add these packages to `dependencies` of the `eslint-config-devextreme` package?

File: themebuilder-scss/src/modules/compile-manager.ts [hunk 10]
```diff
@@ -1,37 +1,35 @@
...
+      const data = await this.compiler.compile(bundle, items, null);
+      let css = data.result.css.toString();
 
-            if(config.makeSwatch) {
-                const preCompiler = new PreCompiler();
-                const swatchSass = preCompiler.createSassForSwatch(config.outColorScheme, css);
-                const swatchResult = await this.compiler.compile(bundle, [], { data: swatchSass });
-                css = swatchResult.result.css;
-            }
+      if (config.makeSwatch) {
```
Comment (San4es) [inline themebuilder-scss/src/modules/compile-manager.ts:17]:
```suggestion
      if(config.makeSwatch) {
```

File: themebuilder-scss/src/modules/commands.ts [hunk 11]
```diff
@@ -1,5 +1,5 @@
 export default {
-    BUILD_THEME: 'build-theme',
-    BUILD_VARS: 'export-theme-vars',
-    BUILD_META: 'export-theme-meta'
+  BUILD_THEME: 'build-theme',
+  BUILD_VARS: 'export-theme-vars',
+  BUILD_META: 'export-theme-meta',
```
Comment (San4es) [inline themebuilder-scss/src/modules/commands.ts:4]:
```suggestion
  BUILD_META: 'export-theme-meta'
```

File: themebuilder-scss/src/metadata/generate.ts [hunk 12]
```diff
@@ -9,16 +10,16 @@ const commentsRegex = /\s*\/\*[\S\s]*?\*\//g;
...
-        process.exit(1);
-    }
-}
+const generate = async (): Promise<void> => {
+  try {
+    const collector = new MetadataCollector();
+    const sourceFiles = collector.readFiles(stylesDirectory, sourceHandler);
+    await MetadataCollector.saveScssFiles(sourceFiles, stylesDestinationDirectory);
+    await collector.saveMetadata(metadataDestinationFile, version.package);
+  } catch (e) {
```
Comment (San4es) [inline themebuilder-scss/src/metadata/generate.ts:19]:
```suggestion
  } catch(e) {
```

File: themebuilder-scss/package.json (hunk 9 above)
Comment (San4es) [inline themebuilder-scss/package.json:42]: @churkin  can we add these packages to `dependencies` of the `eslint-config-devextreme` package?

File: themebuilder-scss/.eslintrc-themebuilder (hunk 8 above)
Comment (San4es) [inline themebuilder-scss/.eslintrc-themebuilder:16]: Let's try to add also:
```
"eslint:recommended"
"plugin:@typescript-eslint/eslint-recommended"
```
?

See: 
https://www.npmjs.com/package/@typescript-eslint/eslint-plugin#recommended-configs

File: themebuilder-scss/package.json (hunk 3 above)
Comment (babich-a) [reply themebuilder-scss/package.json:23]: there is no problem to run 'eslint .' in themebuilder-scss folder. The problem - run 'eslint .' in the root of the repository. Eslint parse .eslintrc in themebuilder-scss folder (yes, it parse .eslintrc file in folder that we include in .eslintignore). So we need to rename .eslintrc in something other.

File: docker-ci.sh (hunk 4 above)
Comment (babich-a) [reply docker-ci.sh:198]: in this case we need to run 'npm i' in themebuilder-scss folder in both tasks - linter and themebuilder tests.

File: themebuilder-scss/package.json (hunk 6 above)
Comment (babich-a) [reply themebuilder-scss/package.json:36]: eslint asks to install it when we use eslint-config-devextreme

File: .eslintignore (hunk 2 above)
Comment (San4es) [reply .eslintignore:6]: See https://github.com/DevExpress/DevExtreme/pull/13036#discussion_r427819392

File: themebuilder-scss/package.json (hunk 3 above)
Comment (San4es) [reply themebuilder-scss/package.json:23]: It looks strange to me:
- how VS Code will process the ThemeBuilder config?
- why root ESLint parses `.eslintrc` in an ignored folder?

Let's ask ESLint devs about this issue? :)

File: docker-ci.sh (hunk 4 above)
Comment (San4es) [reply docker-ci.sh:198]: Got it.

File: themebuilder-scss/package.json (hunk 6 above)
Comment (San4es) [reply themebuilder-scss/package.json:36]: Ok. Let's continue our discussion here:
https://github.com/DevExpress/DevExtreme/pull/13036#discussion_r427837333

File: themebuilder-scss/.eslintrc-themebuilder (hunk 8 above)
Comment (churkin) [reply themebuilder-scss/.eslintrc-themebuilder:16]: Why?

File: themebuilder-scss/package.json (hunk 9 above)
Comment (churkin) [reply themebuilder-scss/package.json:42]: You should not add dependencies to all plugins, used only. It is reason to use peer dependencies.

**Review by churkin — CHANGES_REQUESTED** [review]

File: themebuilder-scss/package.json (hunk 3 above)
Comment (babich-a) [reply themebuilder-scss/package.json:23]: Eslint really parses all .eslintrc, even in .eslintignore https://github.com/eslint/eslint/issues/13147

File: themebuilder-scss/package.json (hunk 9 above)
Comment (San4es) [reply themebuilder-scss/package.json:42]: Got it. Let's do this.

File: themebuilder-scss/.eslintrc-themebuilder (hunk 8 above)
Comment (San4es) [reply themebuilder-scss/.eslintrc-themebuilder:16]: These presets are recommended by authors:
https://www.npmjs.com/package/@typescript-eslint/eslint-plugin#recommended-configs

**Review by churkin — APPROVED** [review]

File: themebuilder-scss/src/modules/compile-manager.ts (hunk 10 above)
Comment (San4es) [reply themebuilder-scss/src/modules/compile-manager.ts:17]: We decided to research and to make a decision outside the context of this PR.

File: themebuilder-scss/src/modules/commands.ts (hunk 11 above)
Comment (San4es) [reply themebuilder-scss/src/modules/commands.ts:4]: See https://github.com/DevExpress/DevExtreme/pull/13036#discussion_r429117137

File: themebuilder-scss/src/metadata/generate.ts (hunk 12 above)
Comment (San4es) [reply themebuilder-scss/src/metadata/generate.ts:19]: See https://github.com/DevExpress/DevExtreme/pull/13036#discussion_r429117137

File: themebuilder-scss/.eslintrc-themebuilder (hunk 5 above)
Comment (San4es) [reply themebuilder-scss/.eslintrc-themebuilder:8]: See https://github.com/DevExpress/DevExtreme/pull/13036#discussion_r429117137

File: themebuilder-scss/.eslintrc-themebuilder (hunk 8 above)
Comment (San4es) [reply themebuilder-scss/.eslintrc-themebuilder:16]: See https://github.com/DevExpress/DevExtreme/pull/13036#discussion_r429117137

File: themebuilder-scss/src/data/bootstrap-metadata/bootstrap-metadata.ts (hunk 7 above)
Comment (San4es) [reply themebuilder-scss/src/data/bootstrap-metadata/bootstrap-metadata.ts:2]: See https://github.com/DevExpress/DevExtreme/pull/13036#discussion_r429117137

**Pull request merged**